Importing the necessary files.

Tokenizer is used to create tokens of words.


In [2]:
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.layers import Embedding,LSTM,Dense
from tensorflow.keras.models import Sequential
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.optimizers import Adam
import pickle
import numpy as np
import os

We open our dataset file and append the lines one by one into a list.

In [11]:
file=open("C:/Krish/Programming/Machine Learning/Files/Data/metamorphosis_clean.txt","r",encoding="utf8")
lines=[]
for i in file:
    lines.append(i)

In [12]:
print("The First Line: ",lines[0])
print("The Last Line: ",lines[1])

The First Line:  One morning, when Gregor Samsa woke from troubled dreams, he found

The Last Line:  himself transformed in his bed into a horrible vermin.  He lay on



In [13]:
lines[0]

'\ufeffOne morning, when Gregor Samsa woke from troubled dreams, he found\n'

In [14]:
lines[1]

'himself transformed in his bed into a horrible vermin.  He lay on\n'

## Cleaning Data

Here we clean our texts by replacing the unnecessary characters like '\n','\r' with a space.?

In [34]:
lines[:10]

['\ufeffOne morning, when Gregor Samsa woke from troubled dreams, he found\n',
 'himself transformed in his bed into a horrible vermin.  He lay on\n',
 'his armour-like back, and if he lifted his head a little he could\n',
 'see his brown belly, slightly domed and divided by arches into stiff\n',
 'sections.  The bedding was hardly able to cover it and seemed ready\n',
 'to slide off any moment.  His many legs, pitifully thin compared\n',
 'with the size of the rest of him, waved about helplessly as he\n',
 'looked.\n',
 '\n',
 '"What\'s happened to me?" he thought.  It wasn\'t a dream.  His room,\n']

In [35]:
data=""

for i in lines:
    data=' '.join(lines)
    
data=data.replace('\n','').replace('\r','').replace('\ufeff','')
data[:500]

'One morning, when Gregor Samsa woke from troubled dreams, he found himself transformed in his bed into a horrible vermin.  He lay on his armour-like back, and if he lifted his head a little he could see his brown belly, slightly domed and divided by arches into stiff sections.  The bedding was hardly able to cover it and seemed ready to slide off any moment.  His many legs, pitifully thin compared with the size of the rest of him, waved about helplessly as he looked.  "What\'s happened to me?" he'

### maketrans()
maketrans()- it isused to replace a character from a set of strings.
maketrans(str1,str2,str3)- str1 the character that needs to be replaced,str2 the character that should be placed, str3 the character that should be deleted.
the output for this would be in ascii number. For eg, maketrans('a','b') will give the output {97:98}.

### translate()
translate() is used to convert those ascii values into characters and make the necessary changes. Here a will be replaced with b after we use the translate function.

In [16]:
strr='a'

translate=str.maketrans('a','b')
strr=strr.translate(translate)

print(translate)
print(strr)

{97: 98}
b


Here we convert all the punctuation marks into a space character. by mapping the ascii values of the punctuaion marks with the ascii value of the space character and then using the translate function to convert the values to characters

In [17]:
import string

translator=str.maketrans(string.punctuation,' '*len(string.punctuation))
new_data=data.translate(translator)

new_data[:500]

'One morning  when Gregor Samsa woke from troubled dreams  he found himself transformed in his bed into a horrible vermin   He lay on his armour like back  and if he lifted his head a little he could see his brown belly  slightly domed and divided by arches into stiff sections   The bedding was hardly able to cover it and seemed ready to slide off any moment   His many legs  pitifully thin compared with the size of the rest of him  waved about helplessly as he looked    What s happened to me   he'

In [40]:
print("Ascii character of space(' ')-",ord(' '))
print("Mapped punctuation characters: ")
print(translator)

Ascii character of space(' ')- 32
Mapped punctuation characters: 
{33: 32, 34: 32, 35: 32, 36: 32, 37: 32, 38: 32, 39: 32, 40: 32, 41: 32, 42: 32, 43: 32, 44: 32, 45: 32, 46: 32, 47: 32, 58: 32, 59: 32, 60: 32, 61: 32, 62: 32, 63: 32, 64: 32, 91: 32, 92: 32, 93: 32, 94: 32, 95: 32, 96: 32, 123: 32, 124: 32, 125: 32, 126: 32}


Here we remove all the words that have been repeated more than once.

In [18]:
z=[]

for i in data.split():
    if i  not in z:
        z.append(i)

data=' '.join(z)
data[:500]

'One morning, when Gregor Samsa woke from troubled dreams, he found himself transformed in his bed into a horrible vermin. He lay on armour-like back, and if lifted head little could see brown belly, slightly domed divided by arches stiff sections. The bedding was hardly able to cover it seemed ready slide off any moment. His many legs, pitifully thin compared with the size of rest him, waved about helplessly as looked. "What\'s happened me?" thought. It wasn\'t dream. room, proper human room altho'

In [19]:
sentence='Hi i am a programmer'

sentence1=sentence.split()
sentence2=''.join(sentence1)
print(sentence2)
sentence=' '.join(sentence1)
print(sentence)

Hiiamaprogrammer
Hi i am a programmer


## Tokenization

We create tokens of all the words, i.e we assign a value to each word 

fit_on_texts: This method is used to update the internal vocabulary of the tokenizer based on the provided text
data. It takes a list of texts as input and updates the tokenizer's internal state to include the words present in the texts.

In [67]:
tokenizer=Tokenizer()
tokenizer.fit_on_texts([data])

pickle.dump(tokenizer,open('tokenizer1.pkl','wb'))

sequence_data=tokenizer.texts_to_sequences([data])[0]
sequence_data[:10]

[52, 139, 55, 14, 93, 935, 28, 1313, 936, 4]

In [45]:
tokenizer.word_index

{'the': 1,
 'to': 2,
 'and': 3,
 'he': 4,
 'his': 5,
 'of': 6,
 'was': 7,
 'it': 8,
 'had': 9,
 'in': 10,
 'that': 11,
 'a': 12,
 'as': 13,
 'gregor': 14,
 'with': 15,
 'she': 16,
 'him': 17,
 'her': 18,
 'would': 19,
 'not': 20,
 'but': 21,
 'at': 22,
 'for': 23,
 'they': 24,
 'on': 25,
 'all': 26,
 'room': 27,
 'from': 28,
 'could': 29,
 'be': 30,
 'out': 31,
 'have': 32,
 'if': 33,
 'there': 34,
 'been': 35,
 "gregor's": 36,
 'so': 37,
 'father': 38,
 'sister': 39,
 'this': 40,
 'now': 41,
 'himself': 42,
 'door': 43,
 'then': 44,
 'back': 45,
 'mother': 46,
 'up': 47,
 'even': 48,
 'into': 49,
 'no': 50,
 'did': 51,
 'one': 52,
 'more': 53,
 'their': 54,
 'when': 55,
 'were': 56,
 'what': 57,
 'about': 58,
 'them': 59,
 'way': 60,
 'only': 61,
 'time': 62,
 'i': 63,
 'by': 64,
 'than': 65,
 'you': 66,
 'just': 67,
 'said': 68,
 'little': 69,
 'any': 70,
 'do': 71,
 'get': 72,
 'other': 73,
 'still': 74,
 'first': 75,
 'or': 76,
 'made': 77,
 'go': 78,
 'some': 79,
 'while': 80,
 's

In [46]:
vocab_size=len(tokenizer.word_index)+1
print(vocab_size)

2617


Here we map one word-token to the next word-token as shown in the output

In [47]:
sequences=[]

for i in range(1,len(sequence_data)):
    words=sequence_data[i-1:i+1]
    sequences.append(words)
    
print("The length of sequences are: ",len(sequences))
sequences=np.array(sequences)
sequences[:10]

The length of sequences are:  22046


array([[  52,  139],
       [ 139,   55],
       [  55,   14],
       [  14,   93],
       [  93,  935],
       [ 935,   28],
       [  28, 1313],
       [1313,  936],
       [ 936,    4],
       [   4,  241]])

Here we divide the sequences into X and y variabels.
X being the current word-token 
y being the next word-token

In [96]:
X=[]
y=[]

for i in sequences:
    X.append(i[0])
    y.append(i[1])
    
X=np.array(X)
y=np.array(y)

In [97]:
print("The data is: ",X[:5])
print("The responses are: ",y[:5])

The data is:  [ 52 139  55  14  93]
The responses are:  [139  55  14  93 935]


to_categorical converts the tokens into an array of 0s and 1s with some length.

Example here we have a list(0,1,2,19) and num_classes=20.
Therefore we will have an array of size 20 for each of the values in the list and
all values in the array will be zero except the position of the items in the list.

In [98]:
a = tf.keras.utils.to_categorical([0, 1, 2, 19], num_classes=20)
print(a)

[[1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]]


We do the same here where our array size= vocab_size(2617) and the value is 1 for each of 
token values

In [99]:
y=to_categorical(y,num_classes=vocab_size)
y[:5]

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

## Creating model

In [100]:
model=Sequential()
model.add(Embedding(vocab_size,10,input_length=1))
model.add(LSTM(1000,return_sequences=True))
model.add(LSTM(1000))
model.add(Dense(1000,activation="relu"))
model.add(Dense(vocab_size,activation="softmax"))

In [101]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_2 (Embedding)     (None, 1, 10)             26170     
                                                                 
 lstm_4 (LSTM)               (None, 1, 1000)           4044000   
                                                                 
 lstm_5 (LSTM)               (None, 1000)              8004000   
                                                                 
 dense_4 (Dense)             (None, 1000)              1001000   
                                                                 
 dense_5 (Dense)             (None, 2617)              2619617   
                                                                 
Total params: 15694787 (59.87 MB)
Trainable params: 15694787 (59.87 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [155]:
from tensorflow.keras.callbacks import ModelCheckpoint
from tensorflow.keras.callbacks import ReduceLROnPlateau
from tensorflow.keras.callbacks import TensorBoard

checkpoint = ModelCheckpoint("nextword1.h5", monitor='loss', verbose=1, save_best_only=True, mode='auto')


reduce=ReduceLROnPlateau(monitor='loss',factor=0.2,patience=3,min_lr=0.0001,verbose=1)

logdir='logsnextword1'
tensorboard_Visualization=TensorBoard(log_dir=logdir)

In [156]:
model.compile(loss="categorical_crossentropy", optimizer=Adam(learning_rate=0.001))

In [158]:
model.fit(X, y, epochs=50, batch_size=64, callbacks=[checkpoint, reduce, tensorboard_Visualization])

Epoch 1/50
345/345 [==============================] - ETA: 0s - loss: 3.8719
Epoch 1: loss improved from 3.90922 to 3.87193, saving model to nextword1.h5
345/345 [==============================] - 81s 233ms/step - loss: 3.8719 - lr: 0.0010
Epoch 2/50
345/345 [==============================] - ETA: 0s - loss: 3.8361
Epoch 2: loss improved from 3.87193 to 3.83605, saving model to nextword1.h5
345/345 [==============================] - 80s 231ms/step - loss: 3.8361 - lr: 0.0010
Epoch 3/50
345/345 [==============================] - ETA: 0s - loss: 3.8027
Epoch 3: loss improved from 3.83605 to 3.80275, saving model to nextword1.h5
345/345 [==============================] - 79s 229ms/step - loss: 3.8027 - lr: 0.0010
Epoch 4/50
345/345 [==============================] - ETA: 0s - loss: 3.7732
Epoch 4: loss improved from 3.80275 to 3.77324, saving model to nextword1.h5
345/345 [==============================] - 78s 227ms/step - loss: 3.7732 - lr: 0.0010
Epoch 5/50
345/345 [====================

Epoch 35/50
345/345 [==============================] - ETA: 0s - loss: 3.3234
Epoch 35: loss improved from 3.33157 to 3.32342, saving model to nextword1.h5
345/345 [==============================] - 80s 231ms/step - loss: 3.3234 - lr: 0.0010
Epoch 36/50
345/345 [==============================] - ETA: 0s - loss: 3.3137
Epoch 36: loss improved from 3.32342 to 3.31368, saving model to nextword1.h5
345/345 [==============================] - 79s 230ms/step - loss: 3.3137 - lr: 0.0010
Epoch 37/50
345/345 [==============================] - ETA: 0s - loss: 3.3112
Epoch 37: loss improved from 3.31368 to 3.31115, saving model to nextword1.h5
345/345 [==============================] - 80s 232ms/step - loss: 3.3112 - lr: 0.0010
Epoch 38/50
345/345 [==============================] - ETA: 0s - loss: 3.3015
Epoch 38: loss improved from 3.31115 to 3.30154, saving model to nextword1.h5
345/345 [==============================] - 80s 231ms/step - loss: 3.3015 - lr: 0.0010
Epoch 39/50
345/345 [===========

In [179]:
text=' what a strenuous'
text=text.split(" ")
text=text[-1]
text=''.join(text)

In [180]:
sequence = tokenizer.texts_to_sequences([text])[0]
sequence = np.array(sequence)
        
preds = model.predict(sequence)

1/1 [==============================] - 0s 9ms/step


In [181]:
preds

array([[0.0000000e+00, 5.1278437e-10, 1.5114620e-11, ..., 0.0000000e+00,
        0.0000000e+00, 0.0000000e+00]], dtype=float32)

In [182]:
predicted_word = ""

for key, value in tokenizer.word_index.items():
    if np.array_equal(value, preds):
        predicted_word = key
        break

In [183]:
predicted_word

''

In [184]:
print(preds)

[[0.0000000e+00 5.1278437e-10 1.5114620e-11 ... 0.0000000e+00
  0.0000000e+00 0.0000000e+00]]


In [185]:
for i in preds:
    print(i)

[0.0000000e+00 5.1278437e-10 1.5114620e-11 ... 0.0000000e+00 0.0000000e+00
 0.0000000e+00]


In [186]:
preds[0]

array([0.0000000e+00, 5.1278437e-10, 1.5114620e-11, ..., 0.0000000e+00,
       0.0000000e+00, 0.0000000e+00], dtype=float32)

In [187]:
sequence = tokenizer.texts_to_sequences([text])[0]
# sequence = np.array(sequence)
        
# preds = model.predict(sequence)

In [188]:
sequence

[1340]

In [189]:
np.array(sequence)

array([1340])

In [190]:
model.predict(np.array(sequence))

1/1 [==============================] - 0s 15ms/step


array([[0.0000000e+00, 5.1278437e-10, 1.5114620e-11, ..., 0.0000000e+00,
        0.0000000e+00, 0.0000000e+00]], dtype=float32)

In [191]:
decoded_labels = np.argmax(preds, axis=1)

In [192]:
decoded_labels

array([1341], dtype=int64)

In [193]:
for key,value in tokenizer.word_index.items():
    if value==decoded_labels:
        predi=key

In [194]:
predi

'career'

Project Reference- Bharath-K3